### ISSUE TITLE
Registering a scheme handler in ST and the OS to open pages from browsers

### ISSUE THREAD

**Issue by jandahl** [tosher/Mediawiker#7]
I was wondering how feasible it would be to have a scheme handler to automatically open a page from the browser, e.g.

mediawiker://en.wikipedia.org/wiki/Sublime_Text

… would open Sublime Text with Mediawiker fetching "Sublime Text" with credentials matched against "en.wikipedia.org" from the settings.

Then I could make a Greasemonkey plugin to add an icon next to a page name for "Edit in Sublime Text"

How doable do you think this would be?

(By the way, I love your plugin)

Kind regards,
Jan

**Comment by DrLulz** [comment]
Will this work on an OS X machine?

**Comment by tosher** [comment]
<pre>
OS X can't to test, haven't test machine :(
</pre>

I think it 99% possible to setup, but can't to try..

**Comment by jandahl** [comment]
Almost working. I think the only thing missing right now is getting the script to work just right, and figuring out how to register a shell script as a handler.

##Script

I grabbed this from the Linux portion of the wiki and created the script `~/bin/mediawiker-handler`

    mkdir -p ~/bin && touch ~/bin/mediawiker-handler && chmod +x ~/bin/mediawiker-handler && cat << EOF >~/bin/mediawiker-handler
    #!/usr/bin/env bash
    /Applications/Sublime\ Text.app/Contents/MacOS/Sublime\ Text -w --command "mediawiker_cli {\"url\": \"\$1\"}"
    EOF

I tested with

~/bin/mediawiker-handler en.wikipedia.org/wiki/Sublime_Text

It 'did' open Sublime Text, but nothing further came of it. The console didn't have anything to offer. The only hint of what's wrong is the "open file" name at the top of the tab, namely `mediawiker_cli {"url"`, which points to the script not properly parsing the URL but instead passing 

## File handler 
I used [these instructions](http://superuser.com/questions/548119/how-do-i-configure-custom-url-handlers-on-os-x) to use [RCDefaulApp](http://www.rubicode.com/Software/RCDefaultApp/)'s command line tool, by running 

    ~/Library/PreferencePanes/RCDefaultApp.prefPane/Contents/Resources/lstool write url mediawiker /Applications/Sublime\ Text.app
URL:mediawiker set to /Applications/Sublime Text.app

That works to 'open' Sublime Text, but I can't point it towards the script I made.

So, to summarize, the script almost works, and the file handler almost works.

**Comment by jandahl** [comment]
Then again, forcing it from the Terminal doesn't really work...

    Applications/Sublime\ Text.app/Contents/MacOS/Sublime\ Text -w --command "mediawiker_cli {"url": "http://en.wikipedia.org/wiki/Sublime_Text"}"

and

    Applications/Sublime\ Text.app/Contents/MacOS/Sublime\ Text -w --command "mediawiker_cli {"url": \"http://en.wikipedia.org/wiki/Sublime_Text\"}"

both produce the same result as the script, so maybe the issue is that ST3 handles the incoming command differently on a Mac.

**Comment by tosher** [comment]
valid final command lines on windows:
<pre>
--command "mediawiker_cli {\"url\": \"mediawiker://site.com/wiki/PageName\"}"
--command "mediawiker_cli {\"url\": \"mediawiker://PageName\"}"
</pre>
Plugin will try to open page on current active wiki.. site.com is not required...

**Comment by tosher** [comment]
https://www.sublimetext.com/docs/3/osx_command_line.html

**Comment by jandahl** [comment]
Ooh! I was using the wrong executable. This works:

    /Applications/Sublime\ Text.app/Contents/SharedSupport/bin/subl --command "mediawiker_cli {\"url\": \"mediawiker://en.wikipedia.org/wiki/Sublime_Text\"}"

That means that the script should be

    #!/usr/bin/env bash
    /Applications/Sublime\ Text.app/Contents/SharedSupport/bin/subl -w --command "mediawiker_cli {\"url\": \"\$1\"}"

I'll try to grok the file handler. Maybe I can even skip the script, which would probably be the best solution all round.

**Comment by tosher** [comment]
Thnx! ..and waiting for updates from you :)

**Comment by DrLulz** [comment]
I got it working using what @jandahl shared:
````
/Applications/Sublime\ Text.app/Contents/SharedSupport/bin/subl --command "mediawiker_cli {\"url\": \"mediawiker://en.wikipedia.org/wiki/Sublime_Text\"}"
```
And then created a custom URL scheme described [here](http://www.macosxautomation.com/applescript/linktrigger/) by making an application.

The applescript I used to create an application is below. Change the domain name to yours.
````
on run
	display dialog "This script applet will respond to webpage links beginning with:" & return & return & "mediawiker://" buttons {"OK"} default button 1 with title "Mediawiker Handler" with icon 1
	tell application "Finder" to update (path to me)
end run


on open location this_URL
	
	set {tid, AppleScript's text item delimiters} to {AppleScript's text item delimiters, {"mediawiker://"}}
	set encoded_name to text item 2 of this_URL
	set AppleScript's text item delimiters to tid
	
	set page_name to my decode_text(encoded_name)
	set _name to my replaceText(page_name, "_", space)
	
	set _cmd to "/Applications/Sublime\\ Text.app/Contents/SharedSupport/bin/subl --command \"mediawiker_cli {\\\"url\\\": \\\"mediawiker://www.YOURDOMAINNAME.com/index.php/" & _name & "\\\"}\""
	
	
	do shell script _cmd
	
end open location

on decode_text(this_text)
	set flag_A to false
	set flag_B to false
	set temp_char to ""
	set the character_list to {}
	repeat with this_char in this_text
		set this_char to the contents of this_char
		if this_char is "%" then
			set flag_A to true
		else if flag_A is true then
			set the temp_char to this_char
			set flag_A to false
			set flag_B to true
		else if flag_B is true then
			set the end of the character_list to my decode_chars(("%" & temp_char & this_char) as string)
			set the temp_char to ""
			set flag_A to false
			set flag_B to false
		else
			set the end of the character_list to this_char
		end if
	end repeat
	return the character_list as string
end decode_text

on decode_chars(these_chars)
	copy these_chars to {indentifying_char, multiplier_char, remainder_char}
	set the hex_list to "123456789ABCDEF"
	if the multiplier_char is in "ABCDEF" then
		set the multiplier_amt to the offset of the multiplier_char in the hex_list
	else
		set the multiplier_amt to the multiplier_char as integer
	end if
	if the remainder_char is in "ABCDEF" then
		set the remainder_amt to the offset of the remainder_char in the hex_list
	else
		set the remainder_amt to the remainder_char as integer
	end if
	set the ASCII_num to (multiplier_amt * 16) + remainder_amt
	return (ASCII character ASCII_num)
end decode_chars


to replaceText(someText, oldItem, newItem)
	set {tempTID, AppleScript's text item delimiters} to {AppleScript's text item delimiters, oldItem}
	try
		set {itemList, AppleScript's text item delimiters} to {text items of someText, newItem}
		set {someText, AppleScript's text item delimiters} to {itemList as text, tempTID}
	on error errorMessage number errorNumber
		set AppleScript's text item delimiters to tempTID
		error errorMessage number errorNumber
	end try
	
	return someText
end replaceText
````

I named my app Mediawiker so my `CFBundleIdentifier` was:
````
<key>CFBundleIdentifier</key>
	<string>com.apple.ScriptEditor.id.MediaWiker</string>
````

I then modified the info.plist adding this above `LSRequiresCarbon`.
````
	<key>CFBundleURLTypes</key>
	<array>
		<dict>
			<key>CFBundleURLName</key>
			<string>Mediawiker</string>
			<key>CFBundleURLSchemes</key>
			<array>
				<string>mediawiker</string>
			</array>
		</dict>
	</array>
	<key>LSRequiresCarbon</key>
````

**Comment by tosher** [comment]
Thanx for the info!

And..
<pre>
Plugin will try to open page on current active wiki.. site.com is not required...
</pre>

**Comment by jandahl** [comment]
I've tested @DrLulz' solution to work, put it in a fork and created a pull request. 

In the mean time, it is found [here](https://github.com/jandahl/Mediawiker)

**Comment by DrLulz** [comment]
@jandahl 

I'm curious, why do you need RCDefaultApp? I didn't need it, what am I missing?

**Comment by jandahl** [comment]
Oops! Errr. I'll. Erm. So that, errr, the user can, erm, more easily, errr, compile and... Oh goshdarnit, I missed that part of your post. I'll just incorporate that bit during the coming week. It's a bit late here. 

Much better than using RCDefaultApp, too.

**Comment by jandahl** [comment]
The power of not wanting to read boring certification guides prevails to have me procrastinate by making the changes before I had the actual time. Making another pull request.
